Qual: Improve hurl script, add to pre-commit, add to win-ci

mdeweerd opened the pull request:
# Qual: Improve hurl script, add to pre-commit, add to win-ci

- Improvements to hurl script (get API key, logging for github actions), make independent from PWD;
- Add hook to pre-commit (manual run, is a reminder on how to run it, facilitates running it);
- Add hurl run to win-ci (enabled through parameter in workflow).

mdeweerd committed:
Qual: Add new pre-commit hook for running hurl tests (manual trigger)

mdeweerd committed:
QUAL: Add HURL test integration for Windows CI

- Add HURL test runner to Windows CI workflow
- Include environment variables for test configuration
- Update test execution step to run HURL tests when enabled

JonBendtsen commented:
> @JonBendtsen I've added hurl to the windows flow.  It's only executed on forks and manual workflow triggers at this time.
> 
> 
> 
> See https://github.com/mdeweerd/dolibarr/actions/runs/19409056135 for some results.
> 
> 
> 
> I updated logToCs to extract the hurl errors.

Great idea, though right now something has changed such that the hurl tests that used to work no longer works :-(


Eventually I wish that hurl (or something similar) would be part of the build process so we would detect possible issues earlier

I even see it as useful for Dolibarr administrators because if they run a hurl test to check their setup and detect changes between upgrades and/or configuration changes then they can better tell their users of the changes with a new version.


JonB

JonB

mdeweerd commented:
When tests are not in ci, they end up no longer working.
The same is valid for the static analyzers.  As long as errors were ignored in files, more of the ignored notices were added.

Now run.sh can run just a few tests and I added a pre-commit hook configuration that runs only changed hurl files (`pre-commit run --hook-stage manual hurl`) - however the file need to be added to run the latest version of the hurl script.


ONe reason for failure in my case seem to be that debug information is added - so hurl would need to ignore debug information somehow:

```
Success api/setup/10_setup_company.hurl (1 request(s) in 1138 ms)
error: Assert body value
   --> api/warehouses/10_warehouses.hurl:42:1
    |
    | GET http://{{hostnport}}/api/index.php/warehouses/0
    | ...
 42 | {"error":{"code":400,"message":"Bad Request: No warehouse with id=0 can exist"}}
    |   -{"error":{"code":400,"message":"Bad Request: No warehouse with id=0 can exist"}}
    |   +{
    |   +    "error": {
    |   +        "code": 400,
    |   +        "message": "Bad Request: No warehouse with id=0 can exist"
    |   +    },
    |   +    "debug": {
    |   +        "source": "api_warehouses.class.php:79 at call stage",
    |   +        "stages": {
    |   +            "success": [
    |   +                "get",
    |   +                "route",
    |   +                "negotiate",
    |   +                "authenticate",
    |   +                "validate"
    |   +            ],
    |   +            "failure": [
    |   +                "call",
    |   +                "message"
    |   +            ]
    |   +        }
    |   +    }
    |   +}
    |
```

JonBendtsen commented:
> When tests are not in ci, they end up no longer working. The same is valid for the static analyzers. As long as errors were ignored in files, more of the ignored notices were added.

yeah, I totally agree, I was just not yet ready to do it my self, and I also did not expect that it would break so fast.

I hoped - and this might be what happened ;-) - that someone with more Dolibarr/development/PHP experience than me would take the initiative and put it into the CI :-D

Some of the things hold me back was that I do not know if the build system starts a Dolibarr, what data it contains, which modules are enabled, ... so I feared ending up enabling something which would always fail - and that is not useful either.


> Now run.sh can run just a few tests and I added a pre-commit hook configuration that runs only changed hurl files (`pre-commit run --hook-stage manual hurl`) - however the file need to be added to run the latest version of the hurl script.

Looks great :-)

> ONe reason for failure in my case seem to be that debug information is added - so hurl would need to ignore debug information somehow:

OR ASSERT that there would be debug information ;-)

> 
> ```
> Success api/setup/10_setup_company.hurl (1 request(s) in 1138 ms)
> error: Assert body value
>    --> api/warehouses/10_warehouses.hurl:42:1
>     |
>     | GET http://{{hostnport}}/api/index.php/warehouses/0
>     | ...
>  42 | {"error":{"code":400,"message":"Bad Request: No warehouse with id=0 can exist"}}
>     |   -{"error":{"code":400,"message":"Bad Request: No warehouse with id=0 can exist"}}
>     |   +{
>     |   +    "error": {
>     |   +        "code": 400,
>     |   +        "message": "Bad Request: No warehouse with id=0 can exist"
>     |   +    },
>     |   +    "debug": {
>     |   +        "source": "api_warehouses.class.php:79 at call stage",
>     |   +        "stages": {
>     |   +            "success": [
>     |   +                "get",
>     |   +                "route",
>     |   +                "negotiate",
>     |   +                "authenticate",
>     |   +                "validate"
>     |   +            ],
>     |   +            "failure": [
>     |   +                "call",
>     |   +                "message"
>     |   +            ]
>     |   +        }
>     |   +    }
>     |   +}
>     |
> ```

JonBendtsen commented:
> I'll let you evolve the tests - you can test the updated script.

If I am the only one creating/writing/fixing/updating/... tests then it will take a while before everything is covered.

Preferably I would want the Dolibarr community to get to the point that if a PR that adds new functionality doesn't come with a working hurl test, then it wouldn't get merged - Possibly we need to have someone help writing the tests, but they are not that hard, and often one can find an example to copy.

I do not think that my Dolibarr gives the debug output that you got in the output you included above.

JonBendtsen commented:
> > If I am the only one creating/writing/fixing/updating/... tests then it will take a while before everything is covered.
> 
> Open-source is generally though there. I decided to help with improving the flow, but not so much with the functionnality. I fixed >20k phan notices, implemented a fast script to check for missing translations keys, setup the win-ci flow, ... . And I have some other opensource projects to attend to.

yeah I can imagine it could be tough, volunteers and such. Dolibarr is so far the only project I contribute to.

> > Preferably I would want the Dolibarr community to get to the point that if a PR that adds new functionality doesn't come with a working hurl test, then it wouldn't get merged - Possibly we need to have someone help writing the tests, but they are not that hard, and often one can find an example to copy.
> 
> I am sure there will be some momentum at some point. More examples and even some slide for eldy to present can be helpful.

yeah maybe I should write such a slide - or possibly show up and being able to answer questions and tell about my experience - I just don't speak french, but I do plan to talk and share at the International DevCamp.

In the https://github.com/Dolibarr/dolibarr/blob/develop/test/hurl/README.md file I did write that we (Dolibarr community) should discuss here in 2025, but I haven't seen any, and I also did not write where we should have that discussion. Do you know of any place where such a discussion has happened?

> > I do not think that my Dolibarr gives the debug output that you got in the output you included above.
> 
> It's also in https://github.com/mdeweerd/dolibarr/actions/runs/19411735332/job/55534085760 and the database is built from scratch using `dev/setup/phpunit/setup_conf.sh` which is used in wndows-ci and close to what is in travis.yml too.

okay. So far I just used hurl tests with my own development Dolibarr with a database restored from my production Dolibarr.

JonBendtsen commented:
> > I do not think that my Dolibarr gives the debug output that you got in the output you included above.
> 
> It's also in https://github.com/mdeweerd/dolibarr/actions/runs/19411735332/job/55534085760 and the database is built from scratch using `dev/setup/phpunit/setup_conf.sh` which is used in wndows-ci and close to what is in travis.yml too.

I think that your GUI tests fails because it is not logged in.

mdeweerd commented:
> I think that your GUI tests fails because it is not logged in.

I think this is better:
https://github.com/mdeweerd/dolibarr/actions/runs/19414098187/job/55539703596

mdeweerd committed:
:art: Modification to debug hurl cookie

mdeweerd committed:
:sparkles: Add verbose mode options to hurl test runner

- Added --verbose and --very-verbose options to control hurl output verbosity
- Updated help message to include new options
- Modified test execution to use the VERBOSE variable
- Added shellcheck disable for SC2086 to handle the VERBOSE variable correctly

mdeweerd committed:
very verbose hurl for login cookie

mdeweerd committed:
Qual: Update hurl test hooks to include usage instructions

The hurl test hooks have been updated to include comments with usage instructions.

mdeweerd commented:
Closing because the integration is stalled
